packages/remix-dev/vite/plugin.ts

```diff
@@ -1492,15 +1492,34 @@ async function handleSpaMode(
   let response = await handler(new Request("http://localhost/"));
   invariant(response.status === 200, "Error generating the index.html file");
 
-  // Write out the index.html file for the SPA
-  let htmlPath = path.join(assetsBuildDirectory, "index.html");
-  await fse.writeFile(htmlPath, await response.text());
-
-  viteConfig.logger.info(
-    "SPA Mode: index.html has been written to your " +
-      colors.bold(path.relative(process.cwd(), assetsBuildDirectory)) +
-      " directory"
-  );
+  let html = await response.text();
+
+  if (response.headers.get("Content-Type") === "application/javascript") {
+    await fse.writeFile(
+      path.join(assetsBuildDirectory, "assets", "index.js"),
+      html
+    );
```

Comment:
If the user returns a `application/javascript` response from the handler, we write it out as JS instead of HTML